### PR TITLE
Sp accumulator

### PR DESCRIPTION
**Pull request by gbalduzz**
Single particle accumulator based on the dnfft(_gpu) class.
Depends on ~#27 #22 #16~.

### CODE REVIEW

**Comment by uhaehner** [comment]
test this please

**Comment by gbalduzz** [comment]
Rebased on master.

**Comment by biddisco** [comment]
There are some problems building this due to some `#include "dca/phys/error_computation_type.hpp"` that doesn't exist, and also some domain definitions that are missing.

**Comment by gbalduzz** [comment]
Ok, I missed one dependency.

**Comment by gbalduzz** [comment]
Rebased on master.

**Comment by biddisco** [comment]
The gpu accumulator test fails on my laptop. 
I will attempt to debug it, but please let me know if there are things I should look at or try first.
```
jb-s76:~/build/dca$ /home/biddisco/build/dca/test/unit/phys/dca_step/cluster_solver/shared_tools/accumulation/sp/sp_accumulator_gpu_test
Running main() from gtest_main.cc
[==========] Running 2 tests from 1 test case.
[----------] Global test environment set-up.
[----------] 2 tests from SpAccumulatorGpuTest
[ RUN      ] SpAccumulatorGpuTest.Accumulate
/home/biddisco/src/dca/test/unit/phys/dca_step/cluster_solver/shared_tools/accumulation/sp/sp_accumulator_gpu_test.cpp:51: Failure
Expected: (5e-7) > (diff.l_inf), actual: 5e-07 vs 2.01123e-05
[  FAILED  ] SpAccumulatorGpuTest.Accumulate (91 ms)
[ RUN      ] SpAccumulatorGpuTest.SumTo
[       OK ] SpAccumulatorGpuTest.SumTo (2 ms)
[----------] 2 tests from SpAccumulatorGpuTest (93 ms total)

[----------] Global test environment tear-down
[==========] 2 tests from 1 test case ran. (93 ms total)
[  PASSED  ] 1 test.
[  FAILED  ] 1 test, listed below:
[  FAILED  ] SpAccumulatorGpuTest.Accumulate

 1 FAILED TEST
```

**Comment by gbalduzz** [comment]
@biddisco I just checked on my laptop and it passes. Which version of cuda and compute capabilities do you have? Can you check if it enters in the fallback atomic add implementation at nclude/dca/linalg/util/atomic_add_cuda.cu.hpp:35? You can just place a precompiler message there.

File: include/dca/phys/dca_step/cluster_solver/shared_tools/accumulation/sp/sp_accumulator.hpp [hunk 1]
```diff
@@ -0,0 +1,215 @@
...
+        const ScalarType t_i = config[i].get_tau();
+        const int delta_r = RDmn::parameter_type::subtract(r_j, r_i);
+        const double scaled_tau = (t_i - t_j) * one_div_two_beta;  // + (i == j) * epsilon;
+
+        const int index = bbr_dmn(b_i, b_j, delta_r);
+        const ScalarType f_val = Ms[s](i, j);
+
+        (*cached_nfft_obj_)[s].accumulate(index, scaled_tau, sign * f_val);
+        if (accumulate_m_sqr_)
+          (*cached_nfft_sqr_obj_)[s].accumulate(index, scaled_tau, sign * f_val * f_val);
```
[the remark this answers is not in the document]
Comment (gbalduzz) [reply include/dca/phys/dca_step/cluster_solver/shared_tools/accumulation/sp/sp_accumulator.hpp]: No, it is not really tested directly. If the class were broken, gpu_trunk would not produce the same results in the integration tests (system_level and integration/ctaux directories). I also used it extensively in CT_INT without a problem. Should I add same test of the indexing?

[the remark this answers is not in the document]
Comment (gbalduzz) [reply include/dca/phys/dca_step/cluster_solver/shared_tools/accumulation/sp/sp_accumulator.hpp]: I will answer myself with c982eda9b7d049ada2d00ab377d4b23f8ad57560.

**Comment by biddisco** [comment]
I'm using `sm_60` compute and the compiler hits the #pragma message inside 
`#if __CUDA_ARCH__ < 600`
```
+-----------------------------------------------------------------------------+
| NVIDIA-SMI 396.54                 Driver Version: 396.54                    |
|-------------------------------+----------------------+----------------------+
| GPU  Name        Persistence-M| Bus-Id        Disp.A | Volatile Uncorr. ECC |
| Fan  Temp  Perf  Pwr:Usage/Cap|         Memory-Usage | GPU-Util  Compute M. |
|===============================+======================+======================|
|   0  GeForce GTX 1060    Off  | 00000000:01:00.0  On |                  N/A |
| N/A   62C    P8     8W /  N/A |    222MiB /  6069MiB |      0%      Default |
+-------------------------------+----------------------+----------------------+
```

**Comment by biddisco** [comment]
So The test should really be 
`#if defined(__CUDA_ARCH__) && __CUDA_ARCH__ < 600`
and I've committed that. It does not fix the test, but at least the check for `__CUDA_ARCH__ < 600` is not triggered by the compiler outside of cuda kernels.

**Comment by PDoakORNL** [comment]
So on the Condo P100  sp_accumulator_gpu_test fails in the same fashion.  This needs to be fixed before we merge.  Why is this architecture dependent?

**Comment by yingwaili** [comment]
It's architecture dependent because earlier CUDA does not have atomics for double

**Comment by gbalduzz** [comment]
This should not be architecture dependent as the fallback implementation with atomic CAS should work. I would need as much information as you could provide me: specifically on which architecture it fails, if dnfft1d_gpu fails as well, if the accumulator test fails on gpu_trunk, and the output files I asked for in slack. Thank you all for the effort!

**Comment by uhaehner** [comment]
retest this please

**Pull request merged**